src/signature-help.jl

```diff
@@ -0,0 +1,325 @@
+            nothing
+        end
+    end
+
+    !isnothing(activeParameter) && (activeParameter -= 1) # shift to 0-based
+    parameters = map(make_paraminfo, params)
+    return SignatureInformation(; label, documentation, parameters, activeParameter)
+end
+
+function cursor_siginfos(mod::Module, ps::JS.ParseStream, b::Int)
+    out = SignatureInformation[]
+    call, after_semicolon = let st0 = JS.build_tree(JL.SyntaxTree, ps; ignore_errors=true)
+        # tolerate one-past-last byte. TODO: go back to closest non-whitespace?
+        bas = byte_ancestors(st0, b-1)
+        i = findfirst(st -> JS.kind(st) === K"call", bas)
+        i === nothing && return out
+
+        # If parents of our call are like (function (where (where ... (call |) ...))),
+        # we're actually in a declaration, and shouldn't show signature help.
+        # Are there other cases this misses?
+        j = i + 1
+        while j + 1 <= lastindex(bas) && kind(bas[j+1]) === K"where"
+            j += 1
+        end
+        j <= lastindex(bas) && kind(bas[j]) === K"function" && return out
+
+        after_semicolon = i > 1 && kind(bas[i-1]) === K"parameters" && b > JS.first_byte(bas[i-1])
+        bas[i], after_semicolon
+    end
+    # TODO: dotcall support
+    JS.numchildren(call) === 0 && return out
+
+    # TODO: We could be calling a local variable.  If it shadows a method, our
+    # ignoring it is misleading.  We need to either know about local variables
+    # in this scope (maybe by caching completion info) or duplicate some work.
+    fn = resolve_property(mod, call[1])
+    !isa(fn, Function) && return out
+
+    ca = CallArgs(call)
+
+    # Influence parameter highlighting by selecting the active argument (which
+    # may be mapped to a parameter in make_siginfo).  If cursor is after all
+    # pos. args and not after semicolon, ask for the next param, which may not
+    # exist.  Otherwise, highlight the param for the arg we're in.
+    #
+    # We don't keep commas---do we want the green node here?
+    active_arg = let no_args = ca.kw_i === 1,
+        past_pos_args = no_args || b > JS.last_byte(ca.args[ca.kw_i - 1]) + 1
+        if past_pos_args && !after_semicolon
+            :next
+        else
+            arg_i = findfirst(a -> JS.first_byte(a) <= b <= JS.last_byte(a) + 1, ca.args)
+            isnothing(arg_i) ? :none : arg_i
+        end
+    end
+
+    for m in methods(fn)
+        if compatible_call(m, ca)
+            push!(out, make_siginfo(m, ca, active_arg))
+        end
+    end
+    return out
+end
+
+"""
+textDocument/signatureHelp is requested when one of the negotiated trigger
+characters is typed.  Eglot (emacs) requests it more frequently.
+"""
+function handle_SignatureHelpRequest(server::Server, msg::SignatureHelpRequest)
+    state = server.state
+    uri = URI(msg.params.textDocument.uri)
+    fi = get_fileinfo(state, uri)
+    mod = find_file_module!(state, uri, msg.params.position)
```

Comment:
Since `find_file_module!` dynamically changes the global state `state.completion_module` based on `pos`,
it is safer to use `find_file_module`. `state.completion_module` is an internal state only for supporting `completionItem/resolve`, so other features should not modify/use it.